enhancements/authentication/direct-oidc-configuration.md

```diff
@@ -0,0 +1,457 @@
+replaces:
+  - "/enhancements/authentication/unsupported-direct-use-of-oidc.md"
+superseded-by: []
+---
+
+# Direct OIDC Configuraition
+
+## Summary
+
+This enhancement describes how to allow configuring OIDC directly as the main
+source of identities in an OpenShift cluster.
+
+## Motivation
+
+In vanilla Kubernetes, it is possible to configure a remote OIDC provider as the
+source of identities. This has made it possible to create solutions that are based
+around it. However, these are incompatible with the current OpenShift authentication
+stack.
+
+We want to both enable the aforementioned solutions, and we want to make it possible
+for users to be able to use their OIDC providers' features to their full extent.
+
+### User Stories
+
+* I am a cluster administrator and I want to be able to use the full potential
+  of my OIDC provider in my OpenShift clusters
+* As an OpenShift user, I want to be able to use ID tokens issued by our company's
+  OIDC provider to query the OpenShift API
+
+### Goals
+
+- Provide an API that exposes kube-apiserver OIDC configuration with possible
+  expansions to accommodate recent "Structured Authentication Configuration"
+  Kubernetes enhancement.
+
+### Non-Goals
+
+- Allow using OIDC directly and keep the original OpenShift authentication
+  stack working.
+
+## Proposal
+
+### Workflow Description
+
+1. a cluster gets created
+2. a cluster administrator requests OAuth client credentials from
+   their company's identity team
+3. the cluster administrator configures the cluster with the OAuth
+   client credentials for use with their company's OIDC provider via
+   a configuration API provided by OpenShift
+4. users interacting with the cluster can now use the company's OIDC
+   provider credentials for authentication
+
+#### Variation and form factor considerations
+
+New API is introduced that could be used across the various flavours of OpenShift
+to configure direct use of OIDC credentials when interacting with OpenShift clusters.
+
+The API proposal also includes validation of the configuration that should be
+implementable on any platform.
+
+The implementation aims on the standalone OCP as that's the most complex
+platform so far. Any derived platform can adopt the pieces that apply to it.
+
+### API Extensions
+
+`authentications.config.openshift.io` is the API that should be getting extended.
+
+Currently, setting the value of `Type` to anything other than `IntegratedOAuth`,
+or an empty string that semantically matches `IntegratedOAuth`, is unsupported.
+
+A new value for the `Type` field is introduced - `OIDC`. Setting this value
+allows the user to configure a new field: "OIDCProviders".
+
+```go
+type Authentication struct {
+    ...
+    // OIDCProviders are OIDC identity providers that can issue tokens
+    // for this cluster
+    // Can only be set if "Type" is set to "OIDC".
```

Comment:
HCP: I'd like to think that this switch logic might make it easier for the OCM to display only relevant fields to configure to distinguish Tech Preview config from any other. Let me know if that's not the case.

Reply:
Does that mean that I cannot influence configuration of the hosted cluster that I'm going to get?

Reply:
As an end service consumer you don't interact with the HC API directly. Layer above have control over which features to expose.